Feature/job scheduling

aatauil opened the pull request.

aatauil committed:
add schema prefix to contants

aatauil committed:
create database-utils to check if virtuoso is running

aatauil committed:
move parseResult util to seperate file under utils folder

Mainly for better seperation. Seems cleaner to have different utils in their own file.

aatauil committed:
npm install cron

aatauil committed:
Add waitForDatabase with scheduleJobs callback

When the app starts or restarts the service will wait until db is up and then call the scheduleJobs function (callback of waitForDatabase)

aatauil committed:
add scheduleJobs logic & constants

aatauil committed:
add /delta-schedule route triggered when new scheduleJob is inserted in db

aatauil committed:
add .gitignore node_modules (needed for development)

aatauil committed:
add deletes getter & getDeletesFor method to Delta class

aatauil committed:
add new scheduled-jobs delta routes

aatauil committed:
cleanup + add cron frequency methods

aatauil committed:
use uuidv4 (see description)

mu-javascript-template still using v1. When generating multiple uuid's at the same time they become identical. Thats a problem as we dont want all different types to have similar uuids, so theirfor using uuid4

aatauil committed:
fix missing sourceUrl & taskOperation

aatauil committed:
fix cron not stopping even when object gets deleted

cecemel commented on app.js:
```diff
@@ -12,10 +14,38 @@ app.use(bodyParser.json({
   res.send('Hello from job-controller');
 });
 
+app.post('/delta-scheduledJob', function(req, res) {
```
Perhaps rename to /delta/scheduled-jobs

cecemel commented on app.js:
```diff
@@ -12,10 +14,38 @@ app.use(bodyParser.json({
+  const scheduledJobToAdd = new Delta(req.body).getInsertsFor('http://www.w3.org/1999/02/22-rdf-syntax-ns#type', 'http://vocab.deri.ie/cogs#ScheduledJob');
+  if (scheduledJobToAdd.length) {
+    console.log(`[DELTA] Started working on adding scheduled-job: ${scheduledJobToAdd}`);
+    addScheduledJob({uri: scheduledJobToAdd[0]});
```
Here best too loop, you might have mutliple jobs coming in

cecemel replied:
Also, use try cath and wait, to allow proper error handling (see implementation of the task-flow)

cecemel commented on app.js:
```diff
@@ -12,10 +14,38 @@ app.use(bodyParser.json({
 });
 
+app.post('/delta-scheduledJob', function(req, res) {
+  const scheduledJobToAdd = new Delta(req.body).getInsertsFor('http://www.w3.org/1999/02/22-rdf-syntax-ns#type', 'http://vocab.deri.ie/cogs#ScheduledJob');
```
I realized too late, that not all information you need might be at hands when delta comes in.  So the checks will need to be more robust.. Can explain later.

cecemel replied:
Also, always first handle deletes than inserts.

cecemel replied:
Extra comment about: https://github.com/lblod/job-controller-service/pull/2#discussion_r705449625
Actually if you check for this one only http://schema.org/repeatFrequency, it will probably work in 99% of the times

cecemel commented on app.js:
```diff
@@ -12,10 +14,38 @@ app.use(bodyParser.json({
+  }
+  res.sendStatus(201);
+});
+
```
Same comments here as above

cecemel commented on lib/delta.js:
```diff
@@ -10,10 +10,32 @@ export class Delta {
+        .map(t => t.subject.value);
+  }
+
+  getInsertsForPredicates(predicate) {
```
Perhaps conisider improving `getInsertsFor` with optional object parameter?

cecemel commented on lib/delta.js:
```diff
@@ -10,10 +10,32 @@ export class Delta {
+      .map(t => t.subject.value);
+  }
+
+  getDeletesForPredicates(predicate) {
```
same

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+    `;
+
+    const scheduledJobData = parseResult(await query(queryScheduledJobs));
+    addScheduledJobsToCron(scheduledJobData);
```
I would separate concerns here, in spirit  the previous code.
you make the loading of the scheduledJob in a separate call. the result is you can manipulate in the caller level.

So in app.js
```
const job = loadScheduledJob()
await schedule(job)
```
Consider making a file scheduled-jobs.js which contains the logic to just CRUD the db for the ScheduledJob

cecemel replied:
Also async/await, lets do try catch as late as possible

cecemel replied:
To fight silent fails

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+
+const CRON_JOBS = {};
+
+async function initScheduledJobs() {
```
perhaps make this a parametrized call, so we can keep the global CRON_JOB in app.js it is easier to find this when discovering the code

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+      return;
+    }
+
+    await addScheduledJobsToCron(scheduledJobsList);
```
make a verbose loop here, so we can be more fine grained in try catching. i.e. if one fails, we log the error

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+  }
+}
+
+async function addScheduledJobsToCron(scheduledJobs) {
```
I would here just keep it top level and make an atomic function from it, i.e. addScheduledJobToCron

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+}
+
+async function addScheduledJobsToCron(scheduledJobs) {
+  scheduledJobs.forEach(job => {
```
not a fan of `.forEach` difficult error handling, doesn't work on async stuff. Prefer for/of loop

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+
+async function createNewJob(scheduledJob) {
+  try {
+    const scheduledJobData = await getScheduledJobData(scheduledJob);
```
See also one of my next comments: make a function that load everything of job in a pojo, so less variations of queries needed.

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+  });
+}
+
+async function createNewJob(scheduledJob) {
```
rename this function, e.g. executeScheduledJob

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+async function createNewJob(scheduledJob) {
+  try {
+    const scheduledJobData = await getScheduledJobData(scheduledJob);
+    const newJobQuery = await constructNewJobQuery(scheduledJobData);
```
Not sure why this has been split.

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+  }
+}
+
+async function getScheduledJobData(scheduledJob) {
```
See one of next comments. I think like loadJob, best to have loadScheduledJob

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+  }
+}
+
+async function constructNewJobQuery(scheduledJob) {
```
suggestion constructJobFromScheduledJob?

cecemel replied:
(with immediate update query) see previous comment.

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+    const jobId = uuid();
+    const jobUri = `http://redpencil.data.gift/id/job/${jobId}`;
+    const remoteDataObjectId = uuid();
+    const remoteDataObjectUri = `http://data.lblod.info/id/remote-data-objects/${remoteDataObjectId}`;
```
This is not generic. I know it belongs to the deep/shallow copy we had. but I would now just stick to the plan of creating a new data container. and copying the contents inside the new container.

cecemel replied:
Partially ignore what I said. You're right, this code is required if we allow users to edit the contents of data container. This has far more deeper repercussions than I had anticipated it would have. And would want to avoid these. The jobs-controller should be agnostic about the contents of the data container. 
And I don't see a simple way right now to combine EDIT of container contents and a jobs-controller that can deep-copy its contents without knowing it (in code) what the contents would be. A way out of this situation, would e.g. a configurable shacl profile,  that the code can use as guide to see what needs to be deep-copied. But this requires further thinking, and will be a bridge to far now. 

So where this this leads to. I guess, simply put: not allowing the user to edit a job (except maybe the title)

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+                dct:modified ${sparqlEscapeDateTime(now)};
+                task:operation ${sparqlEscapeUri(scheduledJob.operation)}.
+
+          ${sparqlEscapeUri(remoteDataObjectUri)} a nfo:RemoteDataObject;
+                mu:uuid ${sparqlEscapeString(remoteDataObjectId)};
+                nie:url ${sparqlEscapeString(scheduledJob.sourceUrl)};
+                adms:status ${sparqlEscapeUri(STATUS_READY_TO_BE_CACHED)};
+                rpioHttp:requestHeader ${sparqlEscapeUri(REQUEST_HEADER)};
+                dct:created ${sparqlEscapeDateTime(now)};
+                dct:modified ${sparqlEscapeDateTime(now)};
+                dct:creator ${sparqlEscapeUri(scheduledJob.uri)}.
+
+          ${sparqlEscapeUri(harvestingCollectionUri)} a hrvst:HarvestingCollection;
+                mu:uuid ${sparqlEscapeString(harvestingCollectionId)};
+                dct:creator ${sparqlEscapeUri(scheduledJob.uri)};
+                dct:hasPart ${sparqlEscapeUri(remoteDataObjectUri)}.
+
+          ${sparqlEscapeUri(dataContainerUri)} a nfo:DataContainer;
+                mu:uuid ${sparqlEscapeString(dataContainerId)};
+                task:hasHarvestingCollection ${sparqlEscapeUri(harvestingCollectionUri)}.
```
So this part should  not be removed

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+      }
+    `;
+    return newJobQuery;
+  } catch(err) {
```
Would suggest; No silent errors, or at least not that deep.

cecemel commented on lib/schedule.js:
```diff
@@ -0,0 +1,224 @@
+
+}
+
+async function updateCronFrequency(schedule) {
```
again, I would be in favour of splitting the concerns.

the schedule.js file is more in charge of the schedling boilerplate.

cecemel commented on constants.js:
```diff
@@ -17,7 +19,15 @@ export const PREFIXES = `
 export const TASK_URI_PREFIX = 'http://redpencil.data.gift/id/task/';
 export const ERROR_URI_PREFIX = 'http://redpencil.data.gift/id/jobs/error/';
+
+export const INPUTCONTAINER_URI_PREFIX = 'http://redpencil.data.gift/id/jobs/error/';
+export const HARVESTINGCOLLECTION_URI_PREFIX = 'http://lblod.data.gift/vocabularies/harvesting/';
+export const REMOTEDATAOBJECT_URI_PREFIX = 'http://redpencil.data.gift/id/jobs/error/';
```
not needed see other comments.

cecemel commented on constants.js:
```diff
@@ -2,10 +2,12 @@ export const STATUS_BUSY = 'http://redpencil.data.gift/id/concept/JobStatus/busy
 export const JOB_TYPE = 'http://vocab.deri.ie/cogs#Job';
 export const TASK_TYPE = 'http://redpencil.data.gift/vocabularies/tasks/Task';
 export const ERROR_TYPE= 'http://open-services.net/ns/core#Error';
+export const REQUEST_HEADER = 'http://data.lblod.info/request-headers/accept/text/html';
```
not needed

cecemel commented on constants.js:
```diff
@@ -17,7 +19,15 @@ export const PREFIXES = `
   PREFIX adms: <http://www.w3.org/ns/adms#>
+  PREFIX schema: <http://schema.org/>
+  PREFIX nfo: <http://www.semanticdesktop.org/ontologies/2007/03/22/nfo#>
+  PREFIX hrvst: <http://lblod.data.gift/vocabularies/harvesting/>
+  PREFIX rpioHttp: <http://redpencil.data.gift/vocabularies/http/>
```
not needed

cecemel reviewed (changes requested):
hey Thanks!
See my comments.


I think the key take away is the comment about the copy of the container. It had way more consequences than I expected.

aatauil committed:
rename endpoint to delta/scheduled-job

aatauil committed:
delete frequency delta & methods ( no frequency edit functionality)

aatauil committed:
remove silent errors

aatauil committed:
top level try/catch + loop over found scheduled-jobs

aatauil committed:
Refactor after review

- Move business logic to app.js
- seperate db logic to lib files
- loop through delta results

aatauil commented:
@cecemel Made the changes you suggested + refactored the code like requested (I tried to make it the same pattern as the code that was already there)

cecemel commented:
Hi, 

the logic got more complicated and specific than I anticipated, and I moved all the extension here to https://github.com/lblod/scheduled-job-controller-service. Hence, will close this PR. And favor https://github.com/lblod/scheduled-job-controller-service
